src/lib/Parser.hs

```diff
@@ -260,6 +263,55 @@ topLet = do
         k = if v == mkName "eff" then EffectRowKind else TypeKind
         uTyKind = WithSrc (Just pos) $ UPrimExpr $ TCExpr k
 
+interfaceDef :: Parser [UDecl]
+interfaceDef = do
+  keyWord InterfaceKW
+  -- TODO: allow multi-parameter type classes
+  interfaceName <- upperName <|> symName
+  (oneAnnBinder, pos) <- withPos annBinder
+  keyWord WhereKW
+  record <- withSrc $ URecordTy <$> interfaceRecordFields ":"
+  let tyCon = UConDef interfaceName $ toNest [oneAnnBinder]
+  let consName = mkInterfaceConsName interfaceName
+  let varTypeName = case oneAnnBinder of
+        Bind v   -> varName v
+        -- This will fail during typechecking.
+        Ignore _ -> mkName "_"
+  let funDefs = mkFunDefs (pos, varTypeName, interfaceName) record
+  return $ (UData tyCon [UConDef consName (toNest [Ignore record])]) : funDefs
+  where
+    mkFunDefs :: (SrcPos, Name, Name) -> UExpr -> [UDecl]
+    mkFunDefs meta recordType
+      | (WithSrc _ (URecordTy (NoExt (LabeledItems items)))) <- recordType =
+          mkFunDefs' $ M.toList items
+      | otherwise = error "this should not be reachable"
+      where
+        mkFunDefs' = fmap (\(name, ty :| []) -> mkOneFunDef meta (name, ty))
+    mkOneFunDef :: (SrcPos, Name, Name) -> (Label, UExpr) -> UDecl
+    mkOneFunDef (pos, varTypeName, interfaceName) (fLabel, fType) =
+      let (p, ann) = lhs
+      in let (ann', rhs') = addImplicitImplicitArgs pos ann rhs
+      in ULet PlainLet (p, ann') rhs'
+      where
+        lhs =
+          let uAnnBinder = Bind $ instanceName :> ns (UApp
+                    (PlainArrow ()) (var interfaceName) (var varTypeName))
+          in (patbb fLabel, Just $ ns $ UPi uAnnBinder ClassArrow fType)
+        rhs =
+          let recordPat' = UPatRecord $ Ext (labeledSingleton fLabel (patbb
+                    fLabel)) $ Just patbi
+          in let caseAlt = UAlt (ns (UPatCon (mkInterfaceConsName interfaceName)
+                    $ toNest [ns recordPat'])) (var (mkName fLabel))
+          in let caseExpr' = UCase (var instanceName) [caseAlt]
+          in ns $ ULam (patbb instanceStr, Nothing) ClassArrow (ns caseExpr')
```

Comment:
Suggestion: if you want to avoid having this long `let in` chains, then there are two things you can do:
(1) A single `let` can bind multiple things, so you don't need to repeat the `in let` part. Just put some whitespace in there.
(2) I think that it is also valid to say
```
something = do
  let x = a
  let y = z
  f x y
```
IIUC that will just run the code in the identity monad?